events/README_ClientVPN.md

```diff
@@ -0,0 +1,56 @@
+# Sample Function
+
+The following is a sample Lambda function that receives a Client VPN connection handler request as an input and then validates the IP address input and checks whether the connection source IP is on the allowed list defined as a map inside the function. If the source IP matches an allowed IP address it allows the access, otherwise an error message is presented to the user. Debug logs are generated to CloudWatch Logs. (Note that anything written to stdout or stderr will be logged as CloudWatch Logs events.)
+
+```go
+import (
+	"fmt"
+	"log"
+	"net"
+
+	"encoding/json"
+
+	"github.com/aws/aws-lambda-go/events"
+	"github.com/aws/aws-lambda-go/lambda"
+)
+
+var (
+	AllowedIPs = map[string]bool{
+		"10.11.12.13": true,
+	}
+)
+
+func handler(request events.ClientVPNConnectionHandlerRequest) (events.ClientVPNConnectionHandlerResponse, error) {
+	requestJson, _ := json.MarshalIndent(request, "", "  ")
+	log.Printf("REQUEST: %s", requestJson)
+
+	sourceIP := request.PublicIP
+	if net.ParseIP(sourceIP) == nil {
+		return events.ClientVPNConnectionHandlerResponse{}, fmt.Errorf("Invalid parameter PublicIP passed in request: %q", sourceIP)
+	}
+
+	log.Printf("SOURCE IP: %q", sourceIP)
+
+	if allowed, ok := AllowedIPs[sourceIP]; ok && allowed {
+		log.Printf("Allowing access from: %q", sourceIP)
+		return events.ClientVPNConnectionHandlerResponse{
+			Allow: true,
+			ErrorMsgOnFailedPostureCompliance: "",
+			PostureComplianceStatuses: []string{},
```

Comment:
small nitpick: do you need to set either of these? The nil value for a string is "" and it might be ok for `PostureComplianceStatuses` to be nil?

Reply:
I've written a test like this to see how I can make this happen:

```go
func TestClientVPNConnectionHandlerResponseConstructorValidation(t *testing.T) {
        // read json from file
        inputJSON, err := ioutil.ReadFile("./testdata/clientvpn-connectionhandler-response.json")
        if err != nil {
                t.Errorf("could not open test file. details: %v", err)
        }

        // construct response
        outputEvent := ClientVPNConnectionHandlerResponse{
                Allow: true,
                SchemaVersion: "v1",
        }

        // serialize to json
        outputJSON, err := json.Marshal(outputEvent)
        if err != nil {
                t.Errorf("could not marshal event. details: %v", err)
        }

        assert.JSONEq(t, string(inputJSON), string(outputJSON))
}
```

for the string value that is easy; I can just drop it and it will be `""`, but the array is not so easy and I need it in there; if I understand https://github.com/golang/go/issues/37711 correctly the choice is either a `func NewClientVPNConnectionHandlerResponse` or a custom marshaller that handles this; both feel like they depart from how the rest of the library is used a bit, although there seems to be functions for DynamoDB attributes as well.

Anything you'd suggest? I somehow feel it's easier to keep this in the sample and hopefully people will use the sample to then end up with the right data structure? The alternative that feels clean as well is the function that creates an "empty" object as it could also create the SchemaVersion as well and we can drop the version from the sample code; that would gain something at least.

What do you think?

Reply:
```go
 ClientVPNConnectionHandlerResponse{
                Allow: true,
                SchemaVersion: "v1",
        }
```

should marshal as:

```json
{
	"allow": true,
	"error-msg-on-failed-posture-compliance": "",
	"posture-compliance-statuses": null,
        "schema-version": "v1",
}
```

The question is: is this ok? Or does it cause failures. Does it need to be:

```json
{
	"allow": true,
	"error-msg-on-failed-posture-compliance": "",
	"posture-compliance-statuses": [],
        "schema-version": "v1",
}
```

If it does then yea, we'll need to keep the:

```go
PostureComplianceStatuses: []string{},
```

Reply:
It tested with

```go
                return events.ClientVPNConnectionHandlerResponse{
                        Allow: true,
                        ErrorMsgOnFailedPostureCompliance: "",
                        SchemaVersion: "v1",
                }, nil
```

and

```go
                return events.ClientVPNConnectionHandlerResponse{
                        Allow: true,
                        SchemaVersion: "v1",
                }, nil
```

and they both fail with an "internal" error. I think the `{}` is indeed required to be returned to ClientVPN.

Would you think a `func NewClientVPNConnectionHandlerResponse` would be good or should we leave the example as-is?

Reply:
Ah, that stinks. It's `[]` not `{}` right?

Reply:
Yes, sorry; it's an empty array. Even in JSON. The behavior with ClientVPN I've tested is that both code snippets above in https://github.com/aws/aws-lambda-go/pull/343#discussion_r545870909 don't result in a successful authentication; whereas if I use the original snippet that initializes an empty array instead of nil it works.

So I think we either need to include this in the sample or provide a function that correctly initializes the response struct.